#	Linear Regression

Housing price prediction using the Boston Housing Dataset by capturing linear relationships between one of the input features and the target value.

In [ ]:
import numpy as np
import pandas as pd
from tensorflow.keras.datasets import boston_housing
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
(X_train, y_train), (X_test, y_test) = boston_housing.load_data()

57026/57026 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step


In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
def evaluate_model(name, y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    r2 = r2_score(y_true, y_pred)

    print(f"\n{name}")
    print(f"MSE: {mse:.4f}")
    print(f"R² Score: {r2:.4f}")

In [ ]:
def simple_linear_regression(X_train, X_test, y_train, y_test):
    X_train_simple = X_train[:, [0]]
    X_test_simple = X_test[:, [0]]

    model = LinearRegression()
    model.fit(X_train_simple, y_train)

    y_pred = model.predict(X_test_simple)
    evaluate_model("Simple Linear Regression", y_test, y_pred)

In [ ]:
simple_linear_regression(X_train, X_test, y_train, y_test)


Simple Linear Regression
MSE: 69.0541
R² Score: 0.1705


The model explains only about 17% of the variation in housing prices (R² = 0.1705) and has a high error of 69.05 (MSE), showing that predictions are not very accurate with a single feature.